presto-main/src/main/java/io/prestosql/operator/GroupedTopNRankAccumulator.java

```diff
@@ -0,0 +1,894 @@
+
+        if (lastHeapNodeIndex == heapRootNodeIndex) {
+            // The root is the last node remaining
+            dropHeapNodePeerGroup(groupId, lastHeapNodeIndex, contextEvictionListener);
+        }
+        else {
+            // Pop the root and insert the last peer group back into the heap to ensure a balanced tree
+            heapPopAndInsert(groupId, lastPeerGroupIndex, lastPeerGroupCount, contextEvictionListener);
+        }
+
+        // peerGroupLookup entry will be updated by definition of inserting the last peer group into a new node
+        heapNodeBuffer.deallocate(lastHeapNodeIndex);
+    }
+
+    /**
+     * Detaches (but does not deallocate) the leaf in the bottom right-most position in the heap.
+     * <p>
+     * Given the fixed insertion order, the bottom right-most leaf will correspond to the last leaf node inserted into
+     * the balanced heap.
+     *
+     * @return leaf node index that was detached from the heap
+     */
+    private long heapDetachLastInsertionLeaf(long groupId)
+    {
+        long heapRootNodeIndex = groupIdToHeapBuffer.getHeapRootNodeIndex(groupId);
+        long heapSize = groupIdToHeapBuffer.getHeapSize(groupId);
+
+        long previousNodeIndex = UNKNOWN_INDEX;
+        HeapTraversal.Child childPosition = null;
+        long currentNodeIndex = heapRootNodeIndex;
+
+        heapTraversal.resetWithPathTo(heapSize);
+        while (!heapTraversal.isTarget()) {
+            previousNodeIndex = currentNodeIndex;
+            childPosition = heapTraversal.nextChild();
+            currentNodeIndex = getChildIndex(currentNodeIndex, childPosition);
+            verify(currentNodeIndex != UNKNOWN_INDEX, "Target node must exist");
+        }
+
+        // Detach the last insertion leaf node, but do not deallocate yet
+        if (previousNodeIndex == UNKNOWN_INDEX) {
+            // Last insertion leaf was the root node
+            groupIdToHeapBuffer.setHeapRootNodeIndex(groupId, UNKNOWN_INDEX);
+            groupIdToHeapBuffer.setHeapValueCount(groupId, 0);
+            groupIdToHeapBuffer.setHeapSize(groupId, 0);
+        }
+        else {
+            setChildIndex(previousNodeIndex, childPosition, UNKNOWN_INDEX);
+            groupIdToHeapBuffer.addHeapValueCount(groupId, -heapNodeBuffer.getPeerGroupCount(currentNodeIndex));
+            groupIdToHeapBuffer.addHeapSize(groupId, -1);
+        }
+
+        return currentNodeIndex;
+    }
+
+    /**
+     * Inserts a new row into the heap for the specified group ID.
+     * <p>
+     * The technique involves traversing the heap from the root to a new bottom left-priority leaf position,
+     * potentially swapping heap nodes along the way to find the proper insertion position for the new row.
+     * Insertions always fill the left child before the right, and fill up an entire heap level before moving to the
+     * next level.
+     */
+    private void heapInsert(long groupId, long newPeerGroupIndex, long newPeerGroupCount)
```

Comment:
I feel like this is probably ok IMO, since it always makes sense to carry the peer group with the count in this context -- especially given how all the other heap methods are set up.